From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alejandro <62246450+Tehin@users.noreply.github.com>
Date: Sun, 12 Jul 2026 16:24:55 -0500
Subject: [PATCH] Minemora Patch


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 37ab5684ece8d6e595ee963607bb5c8fcd6bd81c..5074fe85e453d1d8b5441b0fd9114bff70332cba 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3682,6 +3682,9 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     @Deprecated(since = "1.20.1")
     void removeAdditionalChatCompletions(java.util.Collection<String> completions);
+
+    // MoraPaper
+    void setChatDisabled(boolean value);
     // Paper end - custom chat completions API
 
     // Spigot start
