From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alejandro <62246450+Tehin@users.noreply.github.com>
Date: Sun, 12 Jul 2026 16:24:55 -0500
Subject: [PATCH] Minemora Patch


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 37ab5684ece8d6e595ee963607bb5c8fcd6bd81c..5074fe85e453d1d8b5441b0fd9114bff70332cba 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3682,6 +3682,9 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     @Deprecated(since = "1.20.1")
     void removeAdditionalChatCompletions(java.util.Collection<String> completions);
+
+    // MoraPaper
+    void setChatDisabled(boolean value);
     // Paper end - custom chat completions API
 
     // Spigot start
diff --git a/src/main/java/org/bukkit/event/player/PlayerPreTeleportEvent.java b/src/main/java/org/bukkit/event/player/PlayerPreTeleportEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..f87f99e5b72e03b9d5080dcc708fd900621e1d1c
--- /dev/null
+++ b/src/main/java/org/bukkit/event/player/PlayerPreTeleportEvent.java
@@ -0,0 +1,63 @@
+package org.bukkit.event.player;
+
+import io.papermc.paper.datacomponent.item.consumable.ConsumeEffect;
+import io.papermc.paper.entity.TeleportFlag;
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.Unmodifiable;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * MoraPaper - An event called when the teleport is triggered, before it being canceled
+ * by a passenger, etc.
+ */
+public class PlayerPreTeleportEvent extends PlayerMoveEvent {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    private final Set<TeleportFlag> teleportFlags = new HashSet<>();
+    private final PlayerTeleportEvent.TeleportCause cause;
+
+    @ApiStatus.Internal
+    public PlayerPreTeleportEvent(@NotNull final Player player, @NotNull final Location from, @Nullable final Location to, @NotNull final PlayerTeleportEvent.TeleportCause cause) {
+        super(player, from, to);
+        this.cause = cause;
+    }
+
+    /**
+     * Gets the cause of this teleportation event
+     *
+     * @return Cause of the event
+     */
+    @NotNull
+    public PlayerTeleportEvent.TeleportCause getCause() {
+        return this.cause;
+    }
+
+    public void addFlag(TeleportFlag flag) {
+        this.teleportFlags.add(flag);
+    }
+
+    public Set<TeleportFlag> getFlags() {
+        return this.teleportFlags;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+}
diff --git a/src/main/java/org/bukkit/event/player/PlayerSteerMinecartEvent.java b/src/main/java/org/bukkit/event/player/PlayerSteerMinecartEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..e60179babbc6b3baa2af78e97dae181626392489
--- /dev/null
+++ b/src/main/java/org/bukkit/event/player/PlayerSteerMinecartEvent.java
@@ -0,0 +1,70 @@
+package org.bukkit.event.player;
+
+import org.bukkit.entity.Minecart;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Holds information for player movement events
+ */
+public class PlayerSteerMinecartEvent extends PlayerEvent implements Cancellable {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    private final Minecart vehicle;
+    private boolean cancelled;
+
+    @ApiStatus.Internal
+    public PlayerSteerMinecartEvent(@NotNull final Player player, @NotNull final Minecart vehicle) {
+        super(player);
+
+        this.vehicle = vehicle;
+    }
+
+    /**
+     * Retrieves the vehicle involved in the event.
+     *
+     * @return the vehicle entity associated with the event
+     */
+    public Minecart getVehicle() {
+        return this.vehicle;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * If a move or teleport event is cancelled, the player will be moved or
+     * teleported back to the Location as defined by getFrom(). This will not
+     * fire an event
+     */
+    @Override
+    public boolean isCancelled() {
+        return this.cancelled;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * If a move or teleport event is cancelled, the player will be moved or
+     * teleported back to the Location as defined by getFrom(). This will not
+     * fire an event
+     */
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
