From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alejandro <62246450+Tehin@users.noreply.github.com>
Date: Sun, 12 Jul 2026 16:09:15 -0500
Subject: [PATCH] Minemora Patch


diff --git a/net/minecraft/network/chat/OutgoingChatMessage.java b/net/minecraft/network/chat/OutgoingChatMessage.java
index 259cbcbdd3cd47907629772dc7565212bb434729..61b03a0a05179591a91d7768a068e174f0b81cbf 100644
--- a/net/minecraft/network/chat/OutgoingChatMessage.java
+++ b/net/minecraft/network/chat/OutgoingChatMessage.java
@@ -25,6 +25,9 @@ public interface OutgoingChatMessage {
         }
         @Override
         public void sendToPlayer(final ServerPlayer player, final boolean filtered, final ChatType.Bound chatType, final @org.jspecify.annotations.Nullable Component unsigned) {
+            // MoraPaper
+            if (player.chatDisabled) return;
+
             player.connection.sendDisguisedChatMessage(unsigned != null ? unsigned : this.content, chatType);
             // Paper end
         }
@@ -38,6 +41,9 @@ public interface OutgoingChatMessage {
 
         @Override
         public void sendToPlayer(final ServerPlayer player, final boolean filtered, final ChatType.Bound chatType) {
+            // MoraPaper
+            if (player.chatDisabled) return;
+
             // Paper start
             this.sendToPlayer(player, filtered, chatType, null);
         }
diff --git a/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java b/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
index aaa497096f07f2ef9da62b9760f5c9aa0b0263d3..58cfca970db9b2f045e00cc012741208fbdab5b1 100644
--- a/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
@@ -226,7 +226,7 @@ public class ClientboundPlayerInfoUpdatePacket implements Packet<ClientGamePacke
                 player.getUUID(),
                 player.getGameProfile(),
                 listed, // Paper - Add Listing API for Player
-                player.connection.latency(),
+                player.connection == null ? player.fakePing : player.connection.latency(),
                 player.gameMode(),
                 player.getTabListDisplayName(),
                 player.isModelPartShown(PlayerModelPart.HAT),
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index a7caa1280abef05f4bcecd3dd94843f0d16b0ef0..f7cb7dd5ad6e26e579002cd33c89733e6ed7af3c 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -3370,4 +3370,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         return globalTemptationLookup;
     }
     // Paper end - optimise temptation lookups
+
+    // MoraPaper - Disable dragon
+    public void disableDragon() {
+        this.dragonFight = null;
+        this.paperConfig().entities.spawning.scanForLegacyEnderDragon = false;
+    }
+
 }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index ad5c46db4bbb42aca73831d8865d6b196886ad21..2583a1b9bdad2a789868e3c3fd4b95c192a4d38f 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -452,6 +452,12 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
     private final ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder viewDistanceHolder = new ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder();
 
+    // MoraPaper - Completely disable chat
+    public boolean chatDisabled = false;
+
+    // MoraPaper - Fake ping
+    public int fakePing;
+
     @Override
     public final boolean moonrise$isRealPlayer() {
         return this.isRealPlayer;
@@ -499,6 +505,23 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         // CraftBukkit end
     }
 
+    // MoraPaper - optimized constructor for NPCs
+    public ServerPlayer(MinecraftServer server, ServerLevel level, GameProfile gameProfile) {
+        super(level, gameProfile);
+        this.textFilter = server.createTextFilterForPlayer(this);
+        this.gameMode = server.createGameModeForPlayer(this);
+        this.server = server;
+        this.object = null;
+        // CraftBukkit start
+        this.displayName = this.getScoreboardName();
+        this.adventure$displayName = net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper
+        this.bukkitPickUpLoot = true;
+        this.maxHealthCache = this.getMaxHealth();
+        this.recipeBook = null;
+        this.stats = null;
+        this.advancements = null;
+    }
+
     @Override
     public BlockPos adjustSpawnLocation(final ServerLevel level, final BlockPos spawnSuggestion) {
         CompletableFuture<Vec3> future = PlayerSpawnFinder.findSpawn(level, spawnSuggestion);
@@ -2534,6 +2557,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     public void sendSystemMessage(final Component message, final boolean overlay) {
+        // MoraPaper
+        if (chatDisabled) return;
+
         if (this.acceptsSystemMessages(overlay)) {
             this.connection
                 .send(
@@ -2561,6 +2587,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.sendChatMessage(message, filtered, chatType, null);
     }
     public void sendChatMessage(final OutgoingChatMessage message, final boolean filtered, final ChatType.Bound chatType, @Nullable Component unsigned) {
+        // MoraPaper
+        if (chatDisabled) return;
+
         // Paper end
         if (this.acceptsChatMessages()) {
             message.sendToPlayer(this, filtered, chatType, unsigned); // Paper
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index c521b3e2a5990455ffc8d491c7464a5ecab76656..81c52b1a2fae4e315c436e55f0a05aad613b40af 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2902,6 +2902,10 @@ public class ServerGamePacketListenerImpl
             return;
         }
         // CraftBukkit end
+
+        // MoraPaper
+        if (player.chatDisabled) return;
+
         // Paper start - Ensure that client receives chat packets in the same order that we add into the message signature cache
         synchronized (this.messageSignatureCache) {
         this.send(
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index e074c9a362588fe46e47e599c7b81fdac69c91aa..962135512caf41849cdc0186059316730f2fa302 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3015,6 +3015,11 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return this.equipment.get(slot);
     }
 
+    // MoraPaper - Get all items
+    public net.minecraft.world.entity.EntityEquipment getEquipment() {
+        return this.equipment;
+    }
+
     // Leaf start - Multithreaded tracker
     public boolean hasItemInSlots() {
         return !this.equipment.isEmpty();
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 484348e23b8710d07a6a33817cb80293ee905646..18446c801e0137544813105f13f94f75926c748b 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -139,7 +139,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     protected JumpControl jumpControl;
     private final BodyRotationControl bodyRotationControl;
     protected PathNavigation navigation;
-    protected final GoalSelector goalSelector;
+    protected GoalSelector goalSelector;
     public net.minecraft.world.entity.ai.goal.@Nullable FloatGoal goalFloat; // Paper - Allow nerfed mobs to jump and float
     public GoalSelector targetSelector;
     private @Nullable LivingEntity target;
diff --git a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
index 683a71943a04047ab88c948ee80416c3236dc466..e9ff5cdb34d6e6502764f7d0e12c9964155324bf 100644
--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -379,6 +379,13 @@ public class NewMinecartBehavior extends MinecartBehavior {
         if (this.minecart.getFirstPassenger() instanceof ServerPlayer player) {
             Vec3 moveIntent = player.getLastClientMoveIntent();
             if (moveIntent.lengthSqr() > 0.0) {
+                // MoraPaper Start - Steer minecart event
+                final org.bukkit.event.player.PlayerSteerMinecartEvent event = new org.bukkit.event.player.PlayerSteerMinecartEvent(player.getBukkitEntity(), (org.bukkit.entity.Minecart) this.minecart.getBukkitEntity());
+                event.callEvent();
+
+                if (event.isCancelled()) return deltaMovement;
+                // MoraPaper End
+
                 Vec3 riderMovement = moveIntent.normalize();
                 double ownDist = deltaMovement.horizontalDistanceSqr();
                 if (riderMovement.lengthSqr() > 0.0 && ownDist < 0.01) {
diff --git a/net/minecraft/world/level/levelgen/feature/OreFeature.java b/net/minecraft/world/level/levelgen/feature/OreFeature.java
index f38bdffd7734fab297979d38abd168b2b82e16e3..bb764a3ecc4e93bb566d1e7c6a07fc69638b8193 100644
--- a/net/minecraft/world/level/levelgen/feature/OreFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/OreFeature.java
@@ -19,6 +19,26 @@ public class OreFeature extends Feature<OreConfiguration> {
         super(codec);
     }
 
+    public class PlacedOre {
+        private final BlockPos pos;
+        private final BlockState state;
+
+        public PlacedOre(final BlockPos pos, final BlockState state) {
+            this.pos = pos;
+            this.state = state;
+        }
+
+        public BlockPos getPos() {
+            return pos;
+        }
+
+        public BlockState getState() {
+            return state;
+        }
+    }
+
+    public static java.util.function.Consumer<PlacedOre> LISTENER;
+
     @Override
     public boolean place(final FeaturePlaceContext<OreConfiguration> context) {
         RandomSource random = context.random();
@@ -148,6 +168,11 @@ public class OreFeature extends Feature<OreConfiguration> {
                                                                     sectionRelativeX, sectionRelativeY, sectionRelativeZ, targetState.state, false
                                                                 );
                                                                 placed++;
+
+                                                                if (LISTENER != null) {
+                                                                    LISTENER.accept(new PlacedOre(orePos, targetState.state));
+                                                                }
+
                                                                 break;
                                                             }
                                                         }
