From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alejandro <62246450+Tehin@users.noreply.github.com>
Date: Sun, 12 Jul 2026 16:09:15 -0500
Subject: [PATCH] Minemora Patch


diff --git a/net/minecraft/network/chat/OutgoingChatMessage.java b/net/minecraft/network/chat/OutgoingChatMessage.java
index 259cbcbdd3cd47907629772dc7565212bb434729..61b03a0a05179591a91d7768a068e174f0b81cbf 100644
--- a/net/minecraft/network/chat/OutgoingChatMessage.java
+++ b/net/minecraft/network/chat/OutgoingChatMessage.java
@@ -25,6 +25,9 @@ public interface OutgoingChatMessage {
         }
         @Override
         public void sendToPlayer(final ServerPlayer player, final boolean filtered, final ChatType.Bound chatType, final @org.jspecify.annotations.Nullable Component unsigned) {
+            // MoraPaper
+            if (player.chatDisabled) return;
+
             player.connection.sendDisguisedChatMessage(unsigned != null ? unsigned : this.content, chatType);
             // Paper end
         }
@@ -38,6 +41,9 @@ public interface OutgoingChatMessage {
 
         @Override
         public void sendToPlayer(final ServerPlayer player, final boolean filtered, final ChatType.Bound chatType) {
+            // MoraPaper
+            if (player.chatDisabled) return;
+
             // Paper start
             this.sendToPlayer(player, filtered, chatType, null);
         }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index ad5c46db4bbb42aca73831d8865d6b196886ad21..9c88b7e23a1b8e9096d506c86f67737fb70f7435 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -452,6 +452,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
     private final ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder viewDistanceHolder = new ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder();
 
+    // MoraPaper - Completely disable chat
+    public boolean chatDisabled = false;
+
     @Override
     public final boolean moonrise$isRealPlayer() {
         return this.isRealPlayer;
@@ -499,6 +502,23 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         // CraftBukkit end
     }
 
+    // MoraPaper - optimized constructor for NPCs
+    public ServerPlayer(MinecraftServer server, ServerLevel level, GameProfile gameProfile) {
+        super(level, gameProfile);
+        this.textFilter = server.createTextFilterForPlayer(this);
+        this.gameMode = server.createGameModeForPlayer(this);
+        this.server = server;
+        this.object = null;
+        // CraftBukkit start
+        this.displayName = this.getScoreboardName();
+        this.adventure$displayName = net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper
+        this.bukkitPickUpLoot = true;
+        this.maxHealthCache = this.getMaxHealth();
+        this.recipeBook = null;
+        this.stats = null;
+        this.advancements = null;
+    }
+
     @Override
     public BlockPos adjustSpawnLocation(final ServerLevel level, final BlockPos spawnSuggestion) {
         CompletableFuture<Vec3> future = PlayerSpawnFinder.findSpawn(level, spawnSuggestion);
@@ -2534,6 +2554,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     public void sendSystemMessage(final Component message, final boolean overlay) {
+        // MoraPaper
+        if (chatDisabled) return;
+
         if (this.acceptsSystemMessages(overlay)) {
             this.connection
                 .send(
@@ -2561,6 +2584,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.sendChatMessage(message, filtered, chatType, null);
     }
     public void sendChatMessage(final OutgoingChatMessage message, final boolean filtered, final ChatType.Bound chatType, @Nullable Component unsigned) {
+        // MoraPaper
+        if (chatDisabled) return;
+
         // Paper end
         if (this.acceptsChatMessages()) {
             message.sendToPlayer(this, filtered, chatType, unsigned); // Paper
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index c521b3e2a5990455ffc8d491c7464a5ecab76656..81c52b1a2fae4e315c436e55f0a05aad613b40af 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2902,6 +2902,10 @@ public class ServerGamePacketListenerImpl
             return;
         }
         // CraftBukkit end
+
+        // MoraPaper
+        if (player.chatDisabled) return;
+
         // Paper start - Ensure that client receives chat packets in the same order that we add into the message signature cache
         synchronized (this.messageSignatureCache) {
         this.send(
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 484348e23b8710d07a6a33817cb80293ee905646..18446c801e0137544813105f13f94f75926c748b 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -139,7 +139,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     protected JumpControl jumpControl;
     private final BodyRotationControl bodyRotationControl;
     protected PathNavigation navigation;
-    protected final GoalSelector goalSelector;
+    protected GoalSelector goalSelector;
     public net.minecraft.world.entity.ai.goal.@Nullable FloatGoal goalFloat; // Paper - Allow nerfed mobs to jump and float
     public GoalSelector targetSelector;
     private @Nullable LivingEntity target;
diff --git a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
index 683a71943a04047ab88c948ee80416c3236dc466..e9ff5cdb34d6e6502764f7d0e12c9964155324bf 100644
--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -379,6 +379,13 @@ public class NewMinecartBehavior extends MinecartBehavior {
         if (this.minecart.getFirstPassenger() instanceof ServerPlayer player) {
             Vec3 moveIntent = player.getLastClientMoveIntent();
             if (moveIntent.lengthSqr() > 0.0) {
+                // MoraPaper Start - Steer minecart event
+                final org.bukkit.event.player.PlayerSteerMinecartEvent event = new org.bukkit.event.player.PlayerSteerMinecartEvent(player.getBukkitEntity(), (org.bukkit.entity.Minecart) this.minecart.getBukkitEntity());
+                event.callEvent();
+
+                if (event.isCancelled()) return deltaMovement;
+                // MoraPaper End
+
                 Vec3 riderMovement = moveIntent.normalize();
                 double ownDist = deltaMovement.horizontalDistanceSqr();
                 if (riderMovement.lengthSqr() > 0.0 && ownDist < 0.01) {
